src/common/libflux/reduce.c

```diff
@@ -28,179 +28,204 @@
 #include <czmq.h>
 #include "reduce.h"
 #include "reactor.h"
+#include "info.h"
 
 #include "src/common/libutil/log.h"
 #include "src/common/libutil/shortjson.h"
 #include "src/common/libutil/xzmalloc.h"
 
 
-typedef struct hwm_struct *hwm_t;
-
-struct flux_red_struct {
-    flux_sink_f sinkfn;
-    flux_red_f redfn;
+struct flux_reduce_struct {
+    struct flux_reduce_ops ops;
     void *arg;
-    flux_redstack_t items;
+
+    zlist_t *items; /* set of current items */
+    void *old_item; /* flux_reduce_pop() pops old_item if old_flag is true */
+    bool old_flag;
+    uint32_t rank;
+
     flux_t h;
     int flags;
-    int timeout_msec;
     flux_timer_watcher_t *timer;
+    double timeout;
 
     int last_hwm;
     int cur_hwm;
     int cur_batchnum;
 };
 
-struct flux_redstack_struct {
-    zlist_t *l;
-};
 
-static void timer_cb (flux_t h, flux_timer_watcher_t *w, int revents, void *arg);
+static void flush_current (flux_reduce_t *r);
 
-static bool hwm_flushable (flux_red_t r);
-static bool hwm_valid (flux_red_t r);
-
-static flux_redstack_t redstack_create (void)
-{
-    flux_redstack_t s = xzmalloc (sizeof (*s));
-    if (!(s->l = zlist_new ()))
-        oom();
-    return s;
-}
 
-static void redstack_destroy (flux_redstack_t s)
-{
-    if (s) {
-        zlist_destroy (&s->l);
-        free (s);
-    }
-}
-
-void *flux_redstack_pop (flux_redstack_t s)
-{
-    return zlist_pop (s->l);
-}
-
-void flux_redstack_push (flux_redstack_t s, void *item)
+static void timer_cb (flux_t h, flux_timer_watcher_t *w, int revents, void *arg)
 {
-    if (zlist_push (s->l, item) < 0)
-        oom ();
+    flux_reduce_t *r = arg;
+    flush_current (r);
 }
 
-int flux_redstack_count (flux_redstack_t s)
+flux_reduce_t *flux_reduce_create (flux_t h, struct flux_reduce_ops ops,
+                                   double timeout, void *arg, int flags)
 {
-    return zlist_size (s->l);
-}
+    int arity = 2;
+    uint32_t size = 1;
 
-flux_red_t flux_red_create (flux_t h, flux_sink_f sinkfn, void *arg)
-{
-    flux_red_t r = xzmalloc (sizeof (*r));
-    r->arg = arg;
+    if (!h || ((flags & FLUX_REDUCE_HWMFLUSH) && !ops.itemsize)
```

Comment:
I was trying to allow those callbacks to be NULL if the user didn't for some reason want them.  Arguably this would not be very useful without `op.reduce` or `op.sink`, for example.  I could go either way on this, but I'll lean towards keeping it the way it is unless you or others want to argue more strongly.